ext/src/ruby_api/func.rs

```diff
@@ -198,36 +212,67 @@ impl From<BoxValue<Exception>> for InvokeError {
     }
 }
 
-struct Caller<'a> {
-    inner: &'a RefCell<CallerImpl<'a, StoreData>>,
+pub struct Caller<'a> {
+    inner: UnsafeCell<Option<CallerImpl<'a, StoreData>>>,
 }
 
 impl<'a> Caller<'a> {
-    pub fn store_data(&self) -> Value {
-        self.inner.borrow().data().user_data()
+    pub fn new(caller: CallerImpl<'a, StoreData>) -> Self {
+        Self {
+            inner: UnsafeCell::new(Some(caller)),
+        }
+    }
+    pub fn store_data(&self) -> Result<Value, Error> {
+        self.context().map(|ctx| ctx.data().user_data())
     }
 
-    pub fn export(&self, name: RString) -> Result<Option<Value>, Error> {
-        let mut caller_mut = self.inner.borrow_mut();
+    // Taking a Value as self instead of `&self` to avoid re-wrapping &self in a
+    // typed data object, causing an unnecessary object allocation.
+    pub fn export(rb_self: Value, name: RString) -> Result<Option<Value>, Error> {
+        let caller = rb_self.try_convert::<&Self>()?;
+        let inner = unsafe { &mut *caller.inner.get() }
+            .as_mut()
+            .ok_or_else(|| error!("Caller outlived its Func execution"))?;
+
+        let export = match inner.get_export(unsafe { name.as_str() }?) {
+            Some(export) => export,
+            None => return Ok(None),
+        };
+
+        let store = StoreContextValue::Caller(rb_self);
+        let export: Value = match export.ty(caller.context_mut()?) {
+            ExternType::Func(_) => Func::from_inner(store, export.into_func().unwrap()).into(),
+            ExternType::Memory(_) => {
+                Memory::from_inner(store, export.into_memory().unwrap()).into()
+            }
+            _ => return Ok(None), // Other export types are not supported
+        };
 
-        let ext = caller_mut
-            .get_export(unsafe { name.as_str() }?)
-            .map(|export| match export.ty(caller_mut.as_context_mut()) {
-                ExternType::Func(_func) => {
-                    todo!("Handle externs")
-                }
-                ExternType::Memory(_mem) => {
-                    todo!("Handle externs")
-                }
-                ExternType::Table(_table) => {
-                    todo!("Handle externs")
-                }
-                ExternType::Global(_global) => {
-                    todo!("Handle externs")
-                }
-            });
+        Ok(Some(export))
+    }
+
+    pub fn context(&self) -> Result<StoreContext<StoreData>, Error> {
+        unsafe { (*self.inner.get()).as_ref() }
+            .ok_or_else(|| error!("Caller outlived its Func execution"))
+            .map(|c| c.as_context())
+    }
+
+    pub fn context_mut(&self) -> Result<StoreContextMut<StoreData>, Error> {
```

Comment:
Could we implement [`AsContextMut`](https://docs.rs/wasmtime/latest/src/wasmtime/store/context.rs.html#118) instead?

Reply:
That was my first attempt, but that wasn't possible because we can't get a `&mut self` from a Value, and that's what the trait requires:
>```rust
>fn as_context_mut(&mut self) -> StoreContextMut<'_, Self::Data>;
>```

Unless you can think of a way of working around that?this problem?

Reply:
I cannot! Happy with the current approach.